src/rsh/mod.rs

```diff
@@ -59,51 +59,147 @@ pub fn run(initial_state: State) {
     }
 }
 
+#[derive(PartialEq)]
+enum BuildType {
+    None,
+    Single,
+    Double,
+}
+
+struct ParseResult {
+    result: Vec<String>,
+    completed: bool,
+    build_string: String,
+    build_type: BuildType,
+}
+
 fn parse_args(args: &String) -> Vec<String> {
-    let mut result: Vec<String> = Vec::new();
+    let mut parse_result: ParseResult = ParseResult {
+        result: Vec::new(),
+        completed: false,
+        build_string: String::from(""),
+        build_type: BuildType::None,
+    };
 
     if args.len() == 0 {
-        return result
-    }
-
-    let mut building_string: bool = false;
-    let mut build_string: String = String::from("");
-    for string in args.split_whitespace() {
-        if string.starts_with("\"") {
-            // the string is surrounded by quotes - "word"
-            if string.ends_with("\"") {
-                build_string.push_str(string);
-                result.push(build_string);
-
-                building_string = false;
-                build_string = String::from("");
-            // the string only begins with quote - "word
-            } else {
-                building_string = true;
-
-                build_string.push_str(string);
-                build_string.push(' ');
+        return parse_result.result;
+    }
+
+    let mut argstr = args.clone();
```

Comment:
I'm modifying `argstr` in the loop, but don't want to modify the original `args`as it's used elsewhere in `run`.